### PR TITLE
Recommend HPA

### PR DESCRIPTION
**Pull request by qmhu**

### CODE REVIEW

File: pkg/controller/ehpa/hpa_replicas_controller.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ type HPAReplicasController struct {
 }
 
 func (c *HPAReplicasController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	c.Log.Info("got", "hpa", req.NamespacedName)
+	c.Log.V(4).Info("got", "hpa", req.NamespacedName)
```
Comment (mfanjie) [inline pkg/controller/ehpa/hpa_replicas_controller.go:28]: Got

File: pkg/utils/pod.go [hunk 2]
```diff
@@ -54,3 +55,18 @@ func GetPodCondition(status *v1.PodStatus, conditionType v1.PodConditionType) (i
...
+		for _, c := range pod.Spec.Containers {
+			if containerRequest, ok := c.Resources.Requests[resource]; ok {
+				requests += containerRequest.MilliValue()
+			} else {
+				return 0, fmt.Errorf("missing request for %s", resource)
+			}
+		}
+	}
+	return requests, nil
+}
```
Comment (wenlxie) [inline pkg/utils/pod.go:72]: This function ignore the init container, what's the consideration of this?

Comment (qmhu) [reply pkg/utils/pod.go:72]: Autoscaling is dealing with running pod & containers, init container is not a long running one that should not count in here.

File: pkg/controller/ehpa/effective_hpa_controller.go [hunk 3]
```diff
@@ -20,6 +20,7 @@ import (
 
 	autoscalingapi "github.com/gocrane/api/autoscaling/v1alpha1"
 	predictionapi "github.com/gocrane/api/prediction/v1alpha1"
+
```
Comment (mfanjie) [inline pkg/controller/ehpa/effective_hpa_controller.go:23]: nit: why a new section?

Comment (qmhu) [reply pkg/controller/ehpa/effective_hpa_controller.go:23]: In newest code standard, the import parts are system imports, community imports and crane apis and crane imports.

File: pkg/controller/analytics/analytics_controller.go [hunk 4]
```diff
@@ -254,13 +258,15 @@ func (ac *Controller) createRecommendation(ctx context.Context, a *analysisv1alp
 			OwnerReferences: []metav1.OwnerReference{
 				*newOwnerRef(a),
 			},
-			Labels: id.Labels,
```
Comment (yufeiyu) [inline pkg/controller/analytics/analytics_controller.go:257]: Why remove this?

File: pkg/controller/recommendation/recommendation_controller.go [hunk 5]
```diff
@@ -20,67 +19,86 @@ import (
...
 		}
 		return ctrl.Result{}, err
 	}
 
-	if r.DeletionTimestamp != nil {
+	if recommendation.DeletionTimestamp != nil {
 		// todo stop prediction
-		return ctrl.Result{}, nil
+		return ctrl.Result{
+			RequeueAfter: RsyncPeriod,
```
Comment (yufeiyu) [inline pkg/controller/recommendation/recommendation_controller.go:61]: The cr has been deleted, why we need resync here?

File: pkg/recommend/advisor/ehpa.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	maxCpuUsage := math.SmallestNonzeroFloat64
+	minCpuUsage := math.MaxFloat64
+	// compute historic metric
+	for _, sample := range tsList[0].Samples {
+		if sample.Value > maxCpuUsage {
+			maxCpuUsage = sample.Value
+		} else {
+			minCpuUsage = sample.Value
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:77]: Is this logic correct here?

File: pkg/recommend/advisor/ehpa.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+)
+
+var _ Advisor = &EHPAAdvisor{}
+
+type EHPAAdvisor struct {
+	*types.Context
+}
+
+func (a *EHPAAdvisor) Advise(proposed *types.ProposedRecommendation) error {
+	p := a.Predictors[predictionapi.AlgorithmTypePercentile]
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:28]: DSP

File: pkg/recommend/advisor/ehpa.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	resourceCpu := corev1.ResourceCPU
+	namespace := a.Recommendation.Spec.TargetRef.Namespace
+	if len(namespace) == 0 {
+		namespace = DefaultNamespace
+	}
+	cpuQueryExpr := ResourceToPromQueryExpr(namespace, a.Recommendation.Spec.TargetRef.Name, &resourceCpu)
+
+	klog.V(4).Infof("EHPAAdvisor CpuQuery %s Recommendation %s", cpuQueryExpr, klog.KObj(a.Recommendation))
+	timeNow := time.Now()
+	tsList, err := a.DataSource.QueryTimeSeries(cpuQueryExpr, timeNow.Add(-time.Hour*24*7), timeNow, time.Minute)
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:41]: Why need historical metrics?

File: pkg/recommend/advisor/ehpa.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+				Resource: &autoscalingv2.ResourceMetricSource{
+					Name: resourceCpu,
+					Target: autoscalingv2.MetricTarget{
+						Type:               autoscalingv2.UtilizationMetricType,
+						AverageUtilization: &targetCpuUtilization,
+					},
+				},
+			},
+		},
+		Prediction: &autoscalingapi.Prediction{
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:129]: How can users apply such recommendation with this prediction section?

File: pkg/recommend/advisor/ehpa.go (hunk 8 above)
Comment (qmhu) [reply pkg/recommend/advisor/ehpa.go:41]: When recommend minReplicas and maxReplicas, maybe consider both the history and future is more reliable.

File: pkg/recommend/advisor/ehpa.go (hunk 6 above)
Comment (qmhu) [reply pkg/recommend/advisor/ehpa.go:77]: oh, the logic for minCpuUsage is wrong

File: pkg/controller/analytics/analytics_controller.go (hunk 4 above)
Comment (yufeiyu) [reply pkg/controller/analytics/analytics_controller.go:257]: In case of '云梯'，we get some additional info such as "region", "cluster id" for these labels.

**Pull request merged**